jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -193,38 +194,15 @@ public RetainableByteBuffer acquire(int size, boolean direct)
         RetainedBucket bucket = bucketFor(size, direct);
         if (bucket == null)
             return newRetainableByteBuffer(size, direct, this::removed);
-        Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
 
-        RetainableByteBuffer buffer;
+        Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
         if (entry == null)
-        {
-            Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
-            if (reservedEntry != null)
-            {
-                buffer = newRetainableByteBuffer(bucket._capacity, direct, retainedBuffer ->
-                {
-                    BufferUtil.reset(retainedBuffer.getByteBuffer());
-                    if (!reservedEntry.release())
-                        reservedEntry.remove();
-                });
-
-                // A reserved entry may become old and be evicted before it is enabled.
-                if (reservedEntry.enable(buffer, true))
-                {
-                    if (direct)
-                        _currentDirectMemory.addAndGet(buffer.capacity());
-                    else
-                        _currentHeapMemory.addAndGet(buffer.capacity());
-                    releaseExcessMemory(direct);
-                    return buffer;
-                }
-            }
-            return newRetainableByteBuffer(size, direct, this::removed);
-        }
+            return newRetainableByteBuffer(bucket._capacity, direct, this::reserve);
```

Comment:
Not inserting the buffer into the pool until release has some merit, but a drawback is that we're less accurate tracking the memory used. That's probably not a big deal, though.

Reply:
We always allocate the buffer anyway, so the pool size is really only a limit on idle memory

Reply:
I think we're all in agreement that the pool should only track the memory consumption of idle buffers. So this is a good change.